recipes/cmake/3.x.x/conanfile.py

```diff
@@ -113,6 +114,11 @@ def generate(self):
                 tc.variables["HAVE_POLL_FINE_EXITCODE"] = ''
                 tc.variables["HAVE_POLL_FINE_EXITCODE__TRYRUN_OUTPUT"] = ''
             tc.generate()
+            tc = CMakeDeps(self)
+            # CMake tries to consume openssl::openssl target by default but it fails when checking
+            tc.set_property("openssl", "cmake_find_mode", "module")
```

Comment:
I don't understand the explanation. What is failing exactly?

Reply:
True enough, generation `OpenSSL-config.cmake` fails. When using the config file instead, it tries to link over `openssl::openssl` during those checks supports and fail. Probably trying to consume target instead of variables, or maybe, using target as a library ...

The full log can be read here: https://pastebin.com/EhFFgent

Reply:
Isn't it an occurrence of https://github.com/conan-io/conan/issues/12180 & https://github.com/conan-io/conan/issues/12012?

Reply:
Yes, but not easy to fix I would say. With cmake and non-pattern, projects can do a mess mixing variables and targets names. We would need to provide more properties to allow an custom variable name ...

Reply:
Ok so I would change this comment because it's misleading, and I would refer to conan client issues and PR trying to fix this.

Reply:
Thanks for the references. I ran into this issue as well in https://github.com/conan-io/conan-center-index/pull/16023

Reply:
The underlying reason `try_compile` fails has to do with the the imported targets generated by `CMakeDeps` not having an fall-back configuration when the project is not configured with any. The `cmake_toolchain.cmake` ensures that at least a build configuration is defined (for single-configuration cmake generatos where none is specified) - but the temp project generated by CMake during `try_compile` has the "empty" configuration - and since it doesn't propagate the toolchain, that failsafe is not there and will fail. This can be worked around by defining `CMAKE_TRY_COMPILE_CONFIGURATION` in recipes (by passing the value of `self.settings.build_type`), and this should fix most cases of this.  The lack of a fallback configuration for the imported targets  will be addressed in a future 2.x update, and we may consider in the interim defining `CMAKE_TRY_COMPILE_CONFIGURATION` if we can ensure this doesn't have side effects. We are currently not contemplating the option of turning those variables to return paths instead of imported targets. 

The reason this doesn't work in this particular instance of CMake, has to do with the their invocation of `try_compile` - in this case, this isn't a problem with the _value_ of the variable being an imported target generated by cmakedeps (`openssl::openssl`) but the fact that CMake is not propagating the imported targets _at all_ due to how `try_compile` is invoked: instead of passing the dependency via the `LINK_LIBRARIES` argument (docs [here](https://cmake.org/cmake/help/latest/command/try_compile.html#command:try_compile)) - it is passed with `-DLINK_LIBRARIES`, and as such cmake doesn't re-export the targets as it otherwise would. Chances are, this code in the CMake build scripts predate the modernised interface that propagated the targets.